### DISCUSSION TITLE
Support for polymorphic components?

### DISCUSSION THREAD

**Discussion by EloB** [storybookjs/storybook#19059] [Help]
My code style for React components is polymorphic and prop spreading like below (this is a super simple example without any nested children):
```jsx
const SomeComponent = ({ as: Component = "button", ...rest }) => {
  return <Component {...rest} />;
};
```

It's pretty easy to type this with Typescript like:
```ts
type PolyProps<T extends ElementType, P = {}, A extends string = "as"> = {
  [K in A]?: T;
} & Omit<P, A> & Omit<ComponentProps<T>, A | keyof P>;
```

Then use it like:

```tsx
type Props<T extends ElementType> = PolyProps<T, {
  message?: string,
}>;

const fallback  = "button";

const SomeComponent = <T extends ElementType = typeof fallback>({
  as,
  message,
  ...rest,
}: Props<T>) => {
  useEffect(() => {
    if (message) alert(message);
  }, [message]);
  
  const Component = as || fallback;
  
  return <Component {...rest} />;
};
```

It it possible to support such pattern? I've seen similiar stuff but not really polymorphic requests. When searching for "polymorphic" I only got one result.

**Comment by yannbf** [comment]
Hey @EloB! 👋 

Can you elaborate on what exact support you mean? If you set up a small example that might help us visualize what your expectations are. Thanks!

**Reply by EloB** [reply]
I will when I have time :)

**Reply by EloB** [reply]
But in short it would be good if the autogenerated docs could visulize that it also inherits props from something. Maybe with a link for instance.

I don't know if you ever used Pixi.js but I like that they show extends and a link to what it extended.
![image](https://user-images.githubusercontent.com/476567/188470616-dc675641-96a2-44fc-a65b-915269ddcceb.png)
https://pixijs.download/release/docs/PIXI.Sprite.html

The current Storybook only "document" components that is not "inherited". It would also be really good if the UI could adapt to was entered in the `as` field. My guess is that Storybook doesn't include full IntrinsicAttributes in the build so it might be a bit problematic. A way forward might be to be able to include the most common "interfaces".

**Reply by EloB** [reply]
Made a small gif demo of the benefit of writting like this.

![react-poly-components](https://user-images.githubusercontent.com/476567/188489174-55be355c-78d3-4f53-b58f-573905d0231c.gif)
If you want to try it out yourself:
https://codesandbox.io/s/polymorphic-component-example-8kcgr7?file=/src/App.tsx

**Reply by EloB** [reply]
This also works with custom components. If you pass in `as={MyCustomComponent}`.

**Reply by EloB** [reply]
This is super ugly but something like this (copy/paste with MSPaint)...
![Idea](https://user-images.githubusercontent.com/476567/188493199-211bb9b6-e258-43f1-b5ce-de5eedda2008.png)

Maybe storybook could have a way to include or exclude "IntrinsicAttributes" because of size reasons and make that label very declarative instead.

**Reply by EloB** [reply]
This way of written components is really handy when you create different sections of a Form (as separate components) because each section can then easily be reused. You only have to change the `as` property (that might default to a Fragment or div) to "form" instead and then be able to use normal form props without the need of rewritting a new component.

I know a lot of people hate prop spreading but using Typescript and doing like above I don't see any drawbacks and also 99% of the UI libraries do prop spreading.
